### PR TITLE
Add support for AudioPlayer directives

### PR DESCRIPTION
**Pull request by mattrayner**
This is an in progress branch to prevent duplication of work.
  
something like the following:
```ruby
audio_player = audio_player.play('https://example.com/file.mp3', 'my-token')
directives = [audio_player]
respond('My audio player', directives: directives)
```

----

Actually ends up being:
```ruby
audio_player.play('https://example.com/file.mp3', 'my-token', speech: 'My audio player')
```

### CODE REVIEW

**Comment by mattrayner** [comment]
@sjmog how do you feel about my current approach? Still a WIP of course

File: spec/ralyxa/response_entities/directives/audio/audio_item_spec.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+require_relative '../../../../spec_helper'
+
+RSpec.describe Ralyxa::ResponseEntities::Directives::Audio::AudioItem do
+  subject(:audio_item) { described_class.new(Ralyxa::ResponseEntities::Directives::Audio::Stream.new('http://localhost', 'token')) }
+
+  describe '#initialize' do
+    it 'sets the expected instance variables' do
+      expect(audio_item.instance_variable_get(:@stream)).to be_a(Ralyxa::ResponseEntities::Directives::Audio::Stream)
+    end
+  end
```
Comment (sjmog) [inline spec/ralyxa/response_entities/directives/audio/audio_item_spec.rb]: I don't think this needs to be tested as it's not really part of the interface for this object

File: spec/ralyxa/response_entities/directives/audio/audio_item_spec.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  describe '#to_h' do
+    it 'generates the expected hash' do
+      expect(audio_item.to_h).to eq({ "stream" => { "url" => "http://localhost", "token" => "token", "offsetInMilliseconds" => 0 } })
+    end
+  end
+
+  describe '#self.as_hash' do
+    it 'generates the expected hash' do
+      expect(described_class.as_hash(Ralyxa::ResponseEntities::Directives::Audio::Stream.new('http://localhost', 'token'))).to eq({ "stream" => { "url" => "http://localhost", "token" => "token", "offsetInMilliseconds" => 0 } })
```
Comment (sjmog) [inline spec/ralyxa/response_entities/directives/audio/audio_item_spec.rb]: Can we split this up a bit?

File: spec/ralyxa/response_entities/directives/audio/stream_spec.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    context 'with only the required values' do
+      subject(:stream) { described_class.new('http://localhost', 'my-token') }
+
+      it 'sets the expected instance variables' do
+        expect(stream.instance_variable_get(:@url)).to eq('http://localhost')
+        expect(stream.instance_variable_get(:@token)).to eq('my-token')
+        expect(stream.instance_variable_get(:@offset_in_milliseconds)).to eq(0)
+        expect(stream.instance_variable_get(:@expected_previous_token)).to eq(nil)
+      end
+    end
```
Comment (sjmog) [inline spec/ralyxa/response_entities/directives/audio/stream_spec.rb]: Ditto here (I avoid testing initializers explicitly, prefer to test behaviour dependent on them at a different level of abstraction to the implementation)

File: spec/ralyxa/response_entities/directives/audio_player/clear_queue_spec.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  describe '#initialize' do
+    context 'with default behaviour' do
+      subject(:clear_queue) { described_class.new }
+
+      it 'sets the @behaviour variable' do
+        expect(clear_queue.instance_variable_get(:@behaviour)).to eq('CLEAR_ENQUEUED')
+      end
+    end
```
Comment (sjmog) [inline spec/ralyxa/response_entities/directives/audio_player/clear_queue_spec.rb]: Same here (although I agree the constant tests are a good idea)

File: spec/ralyxa/response_entities/directives/audio_player_spec.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
+require_relative '../../../spec_helper'
+
+RSpec.describe Ralyxa::ResponseEntities::Directives::AudioPlayer do
+  describe 'self.' do
+    describe '#play' do
```
Comment (sjmog) [inline spec/ralyxa/response_entities/directives/audio_player_spec.rb]: This reads a little funny in the documentation output – prefer `.play` if possible!

File: spec/ralyxa_spec.rb [hunk 6]
```diff
@@ -35,6 +35,10 @@
       Ralyxa.validate_requests = 'foo'
       expect(Ralyxa.validate_requests).to eq('foo')
     end
+
+    it 'raises the expected method missing error for a genuinely missing method' do
+      expect{ Ralyxa.foo }.to raise_error(NoMethodError, 'undefined method `foo\' for Ralyxa:Module')
+    end
```
Comment (sjmog) [inline spec/ralyxa_spec.rb]: Oh yes, nice catch :)

**Comment by sjmog** [comment]
I really like it! Great idea to namespace this under directives as it opens up space for alerts and so on.

**Comment by mattrayner** [comment]
Awesome, I'll amend the tests as per your requests and keep in mind the things you've said. I'm going to do some testing and work around receiving the AudioPlayer requests for audio finishing etc and will probably create a podcast player + radio player running on Ralyxa as per the node SDK.

File: spec/ralyxa/response_entities/directives/audio/audio_item_spec.rb (hunk 2 above)
Comment (mattrayner) [reply spec/ralyxa/response_entities/directives/audio/audio_item_spec.rb]: Do you just mean across multiple lines for readability?

**Comment by sjmog** [comment]
That would be really great – if you feel like you're happy with how you build it, can we link it as an exemplar from the README?

**Comment by mattrayner** [comment]
That would be cool - We should re-assess once I've built them 👍

**Comment by mattrayner** [comment]
@sjmog are these amendments what you had in mind?

**Comment by mattrayner** [comment]
@sjmog I've had to make a number go changes to some core modules to add full functionality - including breaking changes to Ralyxa::ResponseEntities::Card - fancy taking a look and okaying?

I'm in the process of building a radio player and will loop you in once finished - the good new is that this looks like the initial functionality we need for AuidPlayer.

Also - I'm coming to MakersAcademy with Parliament as part of the recruitment efforts on Friday 19th - will you be around for a face-to-face? Would be good to put a face to the avatar 😆

File: README.md [hunk 7]
```diff
@@ -148,7 +148,9 @@ end
 
 # Standard card
 intent "SendStandardCard" do
-  standard_card = card("Hello World", "I'm alive!", "https://placehold.it/200")
+  image = { small: "http://placehold.it/720x480", large: "http://placehold.it/1200x800" }
+  standard_card = card("Hello World", "I'm alive!", image)
```
Comment (sjmog) [inline README.md]: Thoughts on using two parameters for image urls (both optional), which might avoid the breaking change?

**Comment by sjmog** [comment]
Added a comment @mattrayner – I'm off on holiday tomorrow so I won't get to see you! Sorry for the delays in the meantime.

File: README.md (hunk 7 above)
Comment (mattrayner) [reply README.md]: I did consider this, but figured that would, in itself still be a breaking change - modifying the functionality within the card of providing the same url for both small and large images?

**Comment by mattrayner** [comment]
@sjmog - this should now be ready to be reviewed - I've created a set of example skills (they work but the repo needs to be polished) - this is an initial example of how they can work

Please note that the examples are a complete spike - before we reference them anywhere I need to add through tests, translations etc.

At the moment they simply duplicate **most** of the functionality from the node examples - you can use them as an example of how I see this code working:

https://github.com/mattrayner/ralyxa-audio-player-examples

File: README.md (hunk 7 above)
Comment (sjmog) [reply README.md]: So it would be

```ruby
standard_card = card("Hello World", "I'm alive!", "http://placehold.it/720x480", "http://placehold.it/1200x800")
```

And `card` would make the final parameter default to the small image if not present?

I'm generally keen to keep the most-used interfaces as 'stringy' as possible. I'm happy with directives to be an array but if we can avoid users having to manually construct hashes, that feels good to me.

**Comment by sjmog** [comment]
Thank you for the examples @mattrayner ! One thing that jumps out is that there's no elegant way to respond with no speech, just close the session. We should implement that (or at least some guidance for whether to use `nil` or an empty string).

I've left a comment regarding the new card interface – I'm open to being convinced one way or the other!

**Comment by sjmog** [comment]
Actually, let me backtrack on that – where you're using `respond(nil, end_session: true)` or `respond("", end_session: true)` what's stopping you from just using `tell()`?

**Comment by mattrayner** [comment]
I hadn't considered using tell - we'd need to use `tell(nil)`, as providing a response of any kind to the audio intents causes an error - you're limited to just sending audio events and nothing else.

Ofcourse, the above is a huge spike so needs extensive testing/readme additions

I'll look at a massive refactor to dry it all up too

File: README.md (hunk 7 above)
Comment (mattrayner) [reply README.md]: Updated as per your request 👍

**Comment by mattrayner** [comment]
Looking back at your message, do you think the addition of directives to tell would be clear?

```ruby
tell(nil, directives: [audio_player.play_later('https://example.com/file.mp3', 'my-token')])
```

**Comment by mattrayner** [comment]
would it make more sense for us to flip `end_session: true` to be the default, with an override for ask?

**Comment by mattrayner** [comment]
@sjmog, have you had a chance to take another look at this? Would love to know your opinion on `end_session`

**Comment by sjmog** [comment]
I'm thinking that it makes more sense to introduce a different high-level interface. `nil`s creep me out and it feels awkward to use `tell`. If you can't send a text response of any kind to some audio intents (like `PlaybackStarted`) , let's create a new interface for doing that rather than hacking `respond`.

I'm thinking something like this (using the [multiple stream example](https://github.com/mattrayner/ralyxa-audio-player-examples/blob/master/multiple-stream/intents/intent_handlers.rb)):

```ruby
# I removed the user stuff for clarity
intent 'PlayAudio' do
  attributes = AttributesHelper.new({})

  episode = AudioData::STREAMS[attributes.play_order[attributes.index]]

  # builds a response with the correct directive
  audio_player.play(
    episode[:url], 
    attributes.index,
    speech: "Playing #{episode[:title]}" # optional parameter
  )
end

intent 'AMAZON.ResumeIntent' do
  user = User.first(user_id: request.user_id)
  attributes = AttributesHelper.new(user.player_attributes)
  episode = AudioData::STREAMS[attributes.play_order[attributes.index]]

  audio_player.play(
   episode[:url], 
   attributes.index, 
   offset_in_milliseconds: attributes.offset_in_milliseconds) # optional parameter
  )
end

intent 'AMAZON.PauseIntent' do
  audio_player.stop
end

intent 'AMAZON.StopIntent' do
  audio_player.stop
end
```

The `audio_player` is made available just like `respond`, `ask`, `tell`, `card` and so on. `#play` and `#stop` should build a response without the user having to do it manually.

**Comment by mattrayner** [comment]
> **TL;DR** I didn't read your code block properly so wrote the below... Thought I'd leave it here because... Well, it took me 10 minutes :joy: - skip to the bottom for 'yep, sounds good'

Well, here's where things get complicated... You **can** provide a textual response along with an audio player directive itself, *if* you are passing it from within an 'IntentRequest' style intent. For example, when triggering the [PlayAudio response](https://github.com/mattrayner/ralyxa-audio-player-examples/blob/master/multiple-stream/intents/intent_handlers.rb#L13) within the multiple-streams example it reads out the episode title.

You **can not**, however send textual content when responding to Amazon's built-in `AudioPlayer.*` directives ([`AudioPlayer.PlaybackNearlyFinished` valid response types](https://developer.amazon.com/docs/custom-skills/audioplayer-interface-reference.html#valid-response-types-3)). These can all be found within [the audio_events.rb file here](https://github.com/mattrayner/ralyxa-audio-player-examples/blob/master/multiple-stream/intents/audio_events.rb), for example.

We could create a new top-level action like `ask`, `respond` etc, but having `audio_player` methods build a response may not work as you've described above 😞.

---

Yep, sounds good... I'll take a look now

**Comment by sjmog** [comment]
Awesome! Which bit was the bit that changed your mind? The optional params hash bit?

**Comment by mattrayner** [comment]
Yeah, I missed the speech param - I also think we should keep the respond
method as t is to allow for flexibility?

**Comment by mattrayner** [comment]
I've done an initial attempt but it looks like I'll have to add a load of extra optionals too (ssml, card, etc.)

**Comment by sjmog** [comment]
Those should all flow straight into `respond`, minus the speech?

I’m fine to bung all the params on there and use AudioPlayer as a builder that filters appropriate params and builds a response with them (using the directives as you set up)

**Comment by mattrayner** [comment]
Unfortunately, as respond is an instance method on Handler, that's not currently the case - it *could* be changed, but I'm not convinced that's a good idea.

Limiting what can be provided along with an audio player directive feels like a good way forward (at least initially) as it should limit unexpected errors caused by users creating combinations of things that are not allowed by the Amazon spec.

This will become even more useful with the other directives (Dialogue, VideoPlayer) as they are even more restrictive in places.

**Comment by mattrayner** [comment]
I've updated both the implementation here, and examples.

I think the creation of comprehensive documentation explaining all the different options for methods would be a good idea - Adding all the various pieces in the README will get unmanageable soon - possibly in a wiki attached to the repo?

It might be advantageous to create some YARD docs for the various ruby classes?

**Comment by sjmog** [comment]
@mattrayner sorry for the delay – a pressured period at work! Is this ready to merge?

**Comment by mattrayner** [comment]
Sorry for the delay my end :joy:

I think it's ready code-wise. The points about documentation etc should probably be converted into issues
